### PR TITLE
Challenge API

### LINKED ISSUE #27: Challenge API
### CHALLENGE API

- [x] Add a Challenege
- [x] Remove a Challenge
- [x] Edit a Challenege

**Difficulty**: Medium 

**Priority**: Medium

### PR DESCRIPTION
**Pull request by khattakdev**
closes #27

### CODE REVIEW

File: server/Controller/challenge.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+  }
+
+  try {
+    let challenge = await challengeDB.findOne({
+      _id: challenge_id,
+      auth0_ref: sub,
+    });
+
+    if (!challenge) {
+      return res.status(402).json({
```
Comment (Drewbi) [inline server/Controller/challenge.js]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/402

I don't think 402 is what you think it is :P

File: server/Controller/challenge.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
+const { challengeDB } = require("../Model/index");
+const Joi = require("@hapi/joi");
+
+exports.getAllChallenges = async (req, res) => {
+  const expertise_id = req.params.expertise_id;
+
+  try {
+    let challenges = await challengeDB.find({ expertise_id: expertise_id });
```
Comment (Drewbi) [inline server/Controller/challenge.js]:
```suggestion
    let challenges = await challengeDB.find({ expertise_id });
```

File: server/Controller/challenge.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
+const { challengeDB } = require("../Model/index");
+const Joi = require("@hapi/joi");
+
+exports.getAllChallenges = async (req, res) => {
```
Comment (Drewbi) [inline server/Controller/challenge.js]:
```suggestion
exports.getExpertiseChallenges = async (req, res) => {
```

Probably a bit more descriptive

File: server/Controller/expertise.js [hunk 4]
```diff
@@ -1,8 +1,23 @@
 const { expertiseDB, userDB } = require("../Model/index");
 const Joi = require("@hapi/joi");
 
+exports.getAllExpertise = async (req, res) => {
```
Comment (Drewbi) [inline server/Controller/expertise.js]:
```suggestion
exports.getUserExpertise = async (req, res) => {
```

File: server/Controller/learning.js [hunk 5]
```diff
@@ -1,19 +1,37 @@
 const { learningDB, userDB } = require("../Model/index");
 const Joi = require("@hapi/joi");
 
+exports.getAllLearnings = async (req, res) => {
```
Comment (Drewbi) [inline server/Controller/learning.js]:
```suggestion
exports.getUserLearnings = async (req, res) => {
```

File: server/Routes/challenge.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+const express = require("express");
+const challengeController = require("../controller/challenge");
+const router = express.Router();
+
+router.post("/register", challengeController.addChallenge);
+router.post("/update", challengeController.updateChallenge);
+router.post("/remove", challengeController.removeChallenge);
+router.get("/:expertise_id", challengeController.getAllChallenges);
```
Comment (Drewbi) [inline server/Routes/challenge.js]:
```suggestion
router.get("/:expertise_id", challengeController.getExpertiseChallenges);
```

File: server/Routes/expertise.js [hunk 7]
```diff
@@ -5,5 +5,6 @@ const router = express.Router();
 router.post("/add", expertiseController.addExpertise);
 router.post("/remove", expertiseController.removeExpertise);
 router.post("/update", expertiseController.updateExpertise);
+router.get("/:user_id", expertiseController.getAllExpertise);
```
Comment (Drewbi) [inline server/Routes/expertise.js]:
```suggestion
router.get("/:user_id", expertiseController.getUserExpertise);
```

File: server/Routes/learning.js [hunk 8]
```diff
@@ -5,5 +5,6 @@ const router = express.Router();
 router.get("/register", userController.addLearning);
 router.get("/remove", userController.removeLearning);
 router.get("/update", userController.updateLearning);
+router.get("/:user_id", userController.getAllLearnings);
```
Comment (Drewbi) [inline server/Routes/learning.js]:
```suggestion
router.get("/:user_id", userController.getUserLearnings);
```

**Review by Drewbi — CHANGES_REQUESTED** [review]
Idm if you leave the controller names the same but probably change the response codes

File: server/Controller/challenge.js (hunk 1 above)
Comment (khattakdev) [reply server/Controller/challenge.js]: Changed to 404

File: server/Routes/challenge.js (hunk 6 above)
Comment (khattakdev) [reply server/Routes/challenge.js]: Updated!

File: server/Routes/expertise.js (hunk 7 above)
Comment (khattakdev) [reply server/Routes/expertise.js]: Updated!

File: server/Routes/learning.js (hunk 8 above)
Comment (khattakdev) [reply server/Routes/learning.js]: Updated!

File: server/controller/challenge.js [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    });
+  } catch (error) {
+    return res.status(400).json({
+      error: error.message,
+    });
+  }
+
+  try {
+    // Find an Expertise with topic [expertise] and auth0_ref [sub]
+    const expertise = expertise.findOne({ topic: expertise, auth0_ref: sub });
```
Comment (anshulrgoyal) [inline server/controller/challenge.js]: it is better to use camelcase `auth0_ref` to `auth0Ref`

File: server/controller/challenge.js [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    description: Joi.string().required(),
+    repo_url: Joi.string(),
+    assets: Joi.string(),
+    cover_photo: Joi.string(),
+  });
+
+  // Schema Validation
+  try {
+    await schema.validateAsync({
+      topic,
```
Comment (anshulrgoyal) [inline server/controller/challenge.js:114]: this validation will fail every time

File: server/controller/challenge.js [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    });
+  } catch (error) {
+    return res.status(400).json({
+      error: error.message,
+    });
+  }
+
+  try {
+    let challenge = await challengeDB.findOne({
```
Comment (anshulrgoyal) [inline server/controller/challenge.js]: use `.update`

File: server/controller/challenge.js [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+      challenge_id,
+    });
+  } catch (error) {
+    return res.status(400).json({
+      error: error.message,
+    });
+  }
+
+  try {
+    let challenge = learningDB.findById(challenge_id);
```
Comment (anshulrgoyal) [inline server/controller/challenge.js]: convert challenge id to object id i think

File: server/controller/expertise.js [hunk 13]
```diff
@@ -71,11 +89,13 @@ exports.updateExpertise = async (req, res) => {
   }
 
   try {
-    const user = userDB.findOne({ authO_ref: nickname });
-    const expertise = expertiseDB.findOne({ topic, user_id: user._id });
+    let expertise = await expertiseDB.findOne({
+      _id: expertise_id,
```
Comment (anshulrgoyal) [inline server/controller/expertise.js]: convert to objectid

**Review by anshulrgoyal — CHANGES_REQUESTED** [review]

File: server/controller/challenge.js (hunk 9 above)
Comment (khattakdev) [reply server/controller/challenge.js]: Forgot to fix this

Comment (khattakdev) [reply server/controller/challenge.js]: Updated!

File: server/controller/challenge.js (hunk 12 above)
Comment (khattakdev) [reply server/controller/challenge.js]: Mongoose converts it automatically.

File: server/controller/expertise.js (hunk 13 above)
Comment (khattakdev) [reply server/controller/expertise.js]: Mongoose converts it automatically if it's a valid ObjectID

File: server/controller/challenge.js (hunk 10 above)
Comment (khattakdev) [reply server/controller/challenge.js:114]: @anshulrgoyal Can you elaborate, please? Why will it fail? I tested it, and it works.

Comment (anshulrgoyal) [reply server/controller/challenge.js:114]: schema has more filed that are required and u are passing only topic

Comment (anshulrgoyal) [reply server/controller/challenge.js:114]: ok when add the comment it had only one field

File: server/controller/challenge.js (hunk 11 above)
Comment (khattakdev) [reply server/controller/challenge.js]: updated!

**Review by anshulrgoyal — APPROVED** [review]

**Pull request merged**